dali/pipeline/operators/decoder/nvjpeg_allocator.h

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_DECODER_NVJPEG_ALLOCATOR_H_
+#define DALI_PIPELINE_OPERATORS_DECODER_NVJPEG_ALLOCATOR_H_
+
+
+#include <cuda_runtime_api.h>
+
+#include <memory>
+#include <mutex>
+#include <numeric>
+#include <string>
+#include <utility>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "dali/common.h"
+#include "dali/error_handling.h"
+#include "dali/util/cuda_utils.h"
+
+namespace dali {
+
+namespace mem {
+class BasicPinnedAllocator {
+ public:
+  static void PreallocateBuffers(size_t element_size_hint, size_t n_elements_hint) {
+    std::lock_guard<std::mutex> l(m_);
+
+    if (element_size_hint_ == 0) {
+      element_size_hint_ = element_size_hint;
+    } else if (element_size_hint_ != element_size_hint) {
+      DALI_FAIL("All instances of nvJPEGDecoder should have the same host_memory_padding.");
+    }
+    free_buffers_pool_.reserve(free_buffers_pool_.size() + n_elements_hint);
+    for (size_t i = 0; i < n_elements_hint; ++i) {
+      void* buffer;
+      CUDA_CALL(cudaHostAlloc(&buffer, element_size_hint, 0));
+      free_buffers_pool_.push_back(buffer);
+    }
+  }
+
+  static void FreeBuffers(size_t n_elements_hint) {
+    std::lock_guard<std::mutex> l(m_);
+    for (size_t i = 0; i < n_elements_hint; ++i) {
+      // Will call std::terminate
+      DALI_ENFORCE(free_buffers_pool_.size() > 0, "Some buffers were not returned by nvJPEG.");
+      CUDA_CALL(cudaFreeHost(free_buffers_pool_.back()));
+      free_buffers_pool_.pop_back();
+    }
+  }
+
+  static int Alloc(void** ptr, size_t size, unsigned int flags) {
+    std::lock_guard<std::mutex> l(m_);
+    // Non managed buffer
+    if (size > element_size_hint_ || free_buffers_pool_.empty()) {
+      return cudaHostAlloc(ptr, size, flags) == cudaSuccess ? 0 : 1;
```

Comment:
this call could be excluded from the mutex lock

Reply:
Done